File: src/infrastructure/storage.postgres/PostgresAggregateStorage.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Threading.Tasks;
using core.fs.Accounts;
using core.Shared;
using Dapper;
using Microsoft.FSharp.Core;
using Newtonsoft.Json;
using Npgsql;
using storage.shared;

namespace storage.postgres
{
    public class PostgresAggregateStorage : IAggregateStorage, IBlobStorage
    {
        private readonly string _cnn;
        private readonly IOutbox _outbox;

        public PostgresAggregateStorage(
            IOutbox outbox,
            string cnn)
        {
            _cnn = cnn;
            _outbox = outbox;
        }

        protected IDbConnection GetConnection()
        {
            var cnn = new NpgsqlConnection(_cnn);
            cnn.Open();
            return cnn; 
        }

        public async Task DeleteAggregates(string entity, UserId userId, IDbTransaction? outsideTransaction = null)
        {
            using var db = outsideTransaction?.Connection ?? GetConnection();
            using var tx = outsideTransaction ?? db.BeginTransaction();
            await db.ExecuteAsync(
                @"DELETE FROM events WHERE entity = :entity AND userId = :userId",
                new { userId = userId.Item, entity }
            );
            tx.Commit();
        }

        public async Task DeleteAggregate(string entity, Guid aggregateId, UserId userId)
        {
            using var db = GetConnection();
            // TODO: no aggregate id in the column, huh?
            // might want to do a migration and add one...
            await db.ExecuteAsync(
                @"DELETE FROM events WHERE entity = :entity AND userId = :userId AND aggregateId = :aggregateId",
                new { userId = userId.Item, entity, aggregateId = aggregateId.ToString() }
            );
        }

        public async Task<IEnumerable<AggregateEvent>> GetEventsAsync(string entity, UserId userId)
        {
            using var db = GetConnection();
            var list = await db.QueryAsync<StoredAggregateEvent>(
                @"select * FROM events WHERE entity = :entity AND userId = :userId ORDER BY version",
                new { entity, userId = userId.Item }
            );

            return list.Select(e => e.Event);
        }
        
        public async Task<IEnumerable<AggregateEvent>> GetEventsAsync(string entity, Guid aggregateId, UserId userId)
        {
            using var db = GetConnection();
            var list = await db.QueryAsync<StoredAggregateEvent>(
                @"select * FROM events WHERE entity = :entity AND userId = :userId AND aggregateId = :aggregateId ORDER BY version",
                new { entity, userId = userId.Item, aggregateId = aggregateId.ToString() }
            );

            return list.Select(e => e.Event);
        }

        private async Task SaveEventsAsyncInternal(IAggregate agg, int fromVersion, string entity, UserId userId,
            IDbTransaction? outsideTransaction = null)
        {
            using var db = outsideTransaction?.Connection ?? GetConnection();
            int version = fromVersion;

            var eventsToBlast = new List<AggregateEvent>();

            using var tx = outsideTransaction ?? db.BeginTransaction();
            foreach (var e in agg.Events.Skip(fromVersion))
            {
                var se = new StoredAggregateEvent
                {
                    Entity = entity,
                    Event = e,
                    Key = e.Id.ToString(),
                    UserId = userId.Item,
                    AggregateId = e.AggregateId.ToString(),
                    Created = DateTimeOffset.UtcNow,
                    Version = ++version
                };

                var query = @"INSERT INTO events (entity, key, aggregateid, userid, created, version, eventjson) VALUES
						(@Entity, @Key, @AggregateId, @UserId, @Created, @Version, @EventJson)";

                await db.ExecuteAsync(query, se);

                eventsToBlast.Add(e);
            }

            await _outbox.AddEvents(eventsToBlast, tx);

            tx.Commit();
        }

        public Task SaveEventsAsync(IAggregate agg, string entity, UserId userId, IDbTransaction? outsideTransaction = null)
        {
            return SaveEventsAsyncInternal(agg, agg.Version, entity, userId, outsideTransaction);
        }

        public Task SaveEventsAsync(IAggregate? oldAggregate, IAggregate newAggregate, string entity, UserId userId, IDbTransaction? outsideTransaction = null)
        {
            return SaveEventsAsyncInternal(newAggregate, oldAggregate?.Version ?? 0, entity, userId, outsideTransaction);
        }

        public async Task DoHealthCheck()
        {
            using var db = GetConnection();
            await db.QueryAsync<int>(@"select 1");
        }

        public async Task<IEnumerable<StoredAggregateEvent>> GetStoredEvents(string entity, UserId userId)
        {
            using var db = GetConnection();
            var list = await db.QueryAsync<StoredAggregateEvent>(
                @"select * FROM events WHERE entity = :entity AND userId = :userId ORDER BY key, version",
                new { entity, userId = userId.Item }
            );

            return list;
        }

        public async Task<FSharpOption<T>> Get<T>(string key)
        {
            using var db = GetConnection();

            var blob = await db.QuerySingleOrDefaultAsync<string>(
                @"select blob FROM blobs WHERE key = :key",
                new { key }
            );

            return string.IsNullOrEmpty(blob) ? 
                FSharpOption<T>.None :
                FSharpOption<T>.Some(JsonConvert.DeserializeObject<T>(blob)!);
        }

        public async Task Save<T>(string key, T t)
        {
            var blob = JsonConvert.SerializeObject(t);

            using var db = GetConnection();

            await db.ExecuteAsync(
                @"INSERT INTO blobs (key, blob, inserted) VALUES (@key, @blob, current_timestamp) ON CONFLICT (key) DO UPDATE SET blob = @blob, inserted = current_timestamp",
                new { key, blob }
            );
        }
    }
}

File: src/infrastructure/storage.postgres.fs/PostgresAggregateStorage.fs
namespace storage.postgres

open System
open System.Data
open System.Linq
open core.fs.Accounts
open core.Shared
open Dapper
open Microsoft.FSharp.Core
open Newtonsoft.Json
open Npgsql
open storage.shared

type PostgresAggregateStorage(outbox: IOutbox, connectionString: string) =
    
    let dataSource = (new NpgsqlDataSourceBuilder(connectionString)).Build()
    
    member _.GetConnection() : IDbConnection =
        dataSource.OpenConnection()
    
    interface IAggregateStorage with
        
        member this.DeleteAggregates(entity: string, userId: UserId, outsideTransaction: IDbTransaction) = 
            task {
                let mutable db: IDbConnection = null
                let mutable tx: IDbTransaction = null
                let ownsConnection = isNull outsideTransaction
                
                try
                    if not (isNull outsideTransaction) then
                        db <- outsideTransaction.Connection
                        tx <- outsideTransaction
                    else
                        db <- this.GetConnection()
                        tx <- db.BeginTransaction()
                    
                    let (UserId id) = userId
                    let sql = "DELETE FROM events WHERE entity = @entity AND userId = @userId"
                    let! _ = db.ExecuteAsync(sql, {| userId = id; entity = entity |})
                    
                    if ownsConnection then
                        tx.Commit()
                with ex ->
                    if ownsConnection && not (isNull tx) then
                        try tx.Rollback() with _ -> ()
                    raise ex
                
                if ownsConnection then
                    if not (isNull tx) then
                        try tx.Dispose() with _ -> ()
                    if not (isNull db) then
                        try db.Dispose() with _ -> ()
            }
        
        member this.DeleteAggregate(entity: string, aggregateId: Guid, userId: UserId) = 
            task {
                use db = this.GetConnection()
                let (UserId id) = userId
                let sql = "DELETE FROM events WHERE entity = @entity AND userId = @userId AND aggregateId = @aggregateId"
                let! _ = db.ExecuteAsync(sql, {| userId = id; entity = entity; aggregateId = aggregateId.ToString() |})
                return ()
            }
        
        member this.GetEventsAsync(entity: string, userId: UserId) = 
            task {
                use db = this.GetConnection()
                let (UserId id) = userId
                let sql = "select * FROM events WHERE entity = @entity AND userId = @userId ORDER BY version"
                let! list = db.QueryAsync<StoredAggregateEvent>(sql, {| entity = entity; userId = id |})
                
                return list.Select(fun e -> StoredAggregateEvent.deserializeEvent(e.EventJson))
            }
        
        member this.GetEventsAsync(entity: string, aggregateId: Guid, userId: UserId) = 
            task {
                use db = this.GetConnection()
                let (UserId id) = userId
                let sql = "select * FROM events WHERE entity = @entity AND userId = @userId AND aggregateId = @aggregateId ORDER BY version"
                let! list = db.QueryAsync<StoredAggregateEvent>(sql, {| entity = entity; userId = id; aggregateId = aggregateId.ToString() |})
                
                return list.Select(fun e -> StoredAggregateEvent.deserializeEvent(e.EventJson))
            }
        
        member this.SaveEventsAsync(agg: IAggregate, entity: string, userId: UserId, outsideTransaction: IDbTransaction) = 
            this.SaveEventsAsyncInternal(agg, agg.Version, entity, userId, outsideTransaction)
        
        member this.SaveEventsAsync(oldAggregate: IAggregate, newAggregate: IAggregate, entity: string, userId: UserId, outsideTransaction: IDbTransaction) = 
            let fromVersion = if isNull oldAggregate then 0 else oldAggregate.Version
            this.SaveEventsAsyncInternal(newAggregate, fromVersion, entity, userId, outsideTransaction)
        
        member this.DoHealthCheck() = 
            task {
                use db = this.GetConnection()
                let! _ = db.QueryAsync<int>("select 1")
                return ()
            }
    
    member private this.SaveEventsAsyncInternal(agg: IAggregate, fromVersion: int, entity: string, userId: UserId, outsideTransaction: IDbTransaction) = 
        task {
            let mutable db: IDbConnection = null
            let mutable tx: IDbTransaction = null
            let ownsConnection = isNull outsideTransaction
            
            try
                if not (isNull outsideTransaction) then
                    db <- outsideTransaction.Connection
                    tx <- outsideTransaction
                else
                    db <- this.GetConnection()
                    tx <- db.BeginTransaction()
                
                let mutable version = fromVersion
                let eventsToBlast = System.Collections.Generic.List<AggregateEvent>()
                let (UserId id) = userId
                
                for e in agg.Events.Skip(fromVersion) do
                    version <- version + 1
                    let se = StoredAggregateEvent.create entity id (e.AggregateId.ToString()) (e.Id.ToString()) DateTimeOffset.UtcNow version e
                    
                    let query = "INSERT INTO events (entity, key, aggregateid, userid, created, version, eventjson) VALUES (@Entity, @Key, @AggregateId, @UserId, @Created, @Version, @EventJson)"
                    
                    let! _ = db.ExecuteAsync(query, se)
                    eventsToBlast.Add(e)
                
                let! result = outbox.AddEvents(eventsToBlast, tx)
                match result with
                | Ok _ -> ()
                | Error e -> failwithf "Failed to add events to outbox: %A" e
                
                if ownsConnection then
                    tx.Commit()
            with ex ->
                if ownsConnection && not (isNull tx) then
                    try tx.Rollback() with _ -> ()
                if ownsConnection then
                    if not (isNull tx) then try tx.Dispose() with _ -> ()
                    if not (isNull db) then try db.Dispose() with _ -> ()
                raise ex
            
            if ownsConnection then  
                if not (isNull tx) then  
                    try tx.Dispose() with _ -> ()  
                if not (isNull db) then  
                    try db.Dispose() with _ -> ()  
        }
    
    member this.GetStoredEvents(entity: string, userId: UserId) = 
        task {
            use db = this.GetConnection()
            let (UserId id) = userId
            let sql = "select * FROM events WHERE entity = @entity AND userId = @userId ORDER BY key, version"
            let! list = db.QueryAsync<StoredAggregateEvent>(sql, {| entity = entity; userId = id |})
            return list
        }
    
    interface IBlobStorage with
        
        member this.Get<'T>(key: string) = 
            task {
                use db = this.GetConnection()
                let sql = "select blob FROM blobs WHERE key = @key"
                let! blob = db.QuerySingleOrDefaultAsync<string>(sql, {| key = key |})
                
                if String.IsNullOrEmpty(blob) then
                    return None
                else
                    let value = JsonConvert.DeserializeObject<'T>(blob)
                    return Some(value)
            }
        
        member this.Save<'T>(key: string, t: 'T) = 
            task {
                let blob = JsonConvert.SerializeObject(t)
                use db = this.GetConnection()
                
                let sql = "INSERT INTO blobs (key, blob, inserted) VALUES (@key, @blob, current_timestamp) ON CONFLICT (key) DO UPDATE SET blob = @blob, inserted = current_timestamp"
                let! _ = db.ExecuteAsync(sql, {| key = key; blob = blob |})
                return ()
            }
